### PR TITLE
feat(boot): one-shot container mode + multi-arch provider image

### PR DESCRIPTION
**Pull request by Falconiere**
Packages toolu-runner as the container image toolu.sh's runner providers (Namespace, Fly) boot: zero-argument ENTRYPOINT, env-only contract, one job, then exit.

## What's in here

**`toolu-runner boot` — env-driven one-shot mode** (`feat(boot)`)
- Reads `TOOLU_JITCONFIG` (GitHub's `encoded_jit_config`, verbatim) straight into `GitHubListener` — no `config.toml`, no credentials store, no `.lock`, no re-mint loop. Never argv: the JIT config would leak into provider dashboards.
- `TOOLU_DEADLINE` (epoch-ms string) is a watchdog backstop, not authority: already-past exits 124 before any network call; mid-job it fires the cancel token (job reports Cancelled to GitHub) then hard-exits after a 30 s grace.
- Exit codes: `0` Success/Skipped, `1` Failure/Cancelled/listener error, `2` env error pre-network, `124` deadline. `GitHubListener::run` now returns `Result<Option<Conclusion>, _>` so the code can reflect the job outcome; `run` / `--once` semantics unchanged.

**rustls everywhere** (`build(wire)`)
- `tokio-tungstenite` flipped from `native-tls` to `rustls-tls-native-roots` — the workspace's sole OpenSSL pull, with zero code referencing the backend. `openssl-sys` is gone from the lockfile; the image ships no libssl.

**Image + publish workflow** (`ci(image)`)
- Multi-stage Dockerfile: `rust:1.94.1-bookworm` → `debian:bookworm-slim`; bash/git/ca-certificates + curl/sudo/tar/gzip/zstd/xz/unzip/jq; Node 20.18.3 + 24.0.2 pre-seeded (ephemeral containers never warm a lazy cache), checksum-verified at build time.
- Multi-arch `linux/amd64` + `linux/arm64` (Apple-Silicon local runs) behind one manifest-list digest; providers keep resolving amd64 from the same pin.
- `release-image.yml`: on tag `v*` — gate → native per-arch builds (no QEMU Rust) → digest-pushed legs → `imagetools` merge → prints the pinnable `ghcr.io/<owner>/toolu-ghrunner@sha256:…` ref in the job summary. PRs touching image inputs build without pushing.
- Docker stays unsupported and documented (`docker://` steps already fail cleanly step-level; no DinD).
- Docs: `docs/container-image.md` (env contract, exit codes, per-provider limits) + README/CLAUDE.md/architecture updates.

## Verification

- fmt / clippy `-D warnings` / 453 tests / protocol-purity guard — all green.
- Exit paths pinned end-to-end through the real subprocess against a real-crypto wiremock broker: Success→0, Failure→1, mid-job watchdog→124, cancel-no-job→`Ok(None)`.
- Container-level: zero-arg run → clear exit-2 naming `TOOLU_JITCONFIG`; past deadline under `--network none` → 124 in ~1 s; both Node seeds execute; binary links no libssl.
- Live smoke against production GitHub: real JIT config minted, container authenticated (RSA→JWT→OAuth2 over rustls), created a broker session, long-polled, exited 124 at the deadline; ephemeral runner cleaned up. The live-log WebSocket path is exercised by the existing live workflows.

## Follow-ups (not this PR)

- First `v*` tag publishes the image; one-time flip of the GHCR package to public (Fly can't pull private registries).
- toolu.sh's smoke-script example names `ghcr.io/toolu/ghrunner`; the real ref is `ghcr.io/falconiere/toolu-ghrunner` — one-line doc fix on that side.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: crates/toolu-runner/src/boot_cmd.rs [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+      () = tokio::time::sleep(remaining) => {},
+    }
+
+    deadline_hit.store(true, Ordering::SeqCst);
+    tracing::error!("TOOLU_DEADLINE reached; cancelling the in-flight job");
+    cancel.cancel();
+
+    tokio::time::sleep(WATCHDOG_GRACE).await;
+    tracing::error!("deadline grace period elapsed without a clean exit; hard-exiting 124");
+    std::process::exit(124);
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:311]: The hard exit is the documented last-resort backstop, not the normal path. At the deadline the watchdog first fires the shared CancellationToken; the graceful path cancels the step, reports Cancelled via /completejob, returns through run_listener_once (which itself yields 124), and main exits — pinned by boot_cmd_e2e_test.rs::boot_exits_124_when_the_deadline_watchdog_fires_mid_job (observed ~7s; grace is 30s), which as of 684490b also asserts the /completejob report landed. std::process::exit(124) fires only when that graceful shutdown has hung for 30s — and a hung flush/teardown is exactly what a deadline backstop must not wait on ('exit only after all async work has been awaited' cannot terminate a hung job by construction; the provider force-kills at its own deadline regardless, per docs/container-image.md). The two 124 paths don't race destructively: same exit code, and the hard path trails the graceful one by 30s.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+/// long enough for a graceful cancel to report `Cancelled` to GitHub, short
+/// enough that a provider's own deadline enforcement never races us.
+const WATCHDOG_GRACE: Duration = Duration::from_secs(30);
+
+/// `boot`: read `TOOLU_JITCONFIG` / `TOOLU_DEADLINE`, run one JIT lifecycle,
+/// return the process exit code: `0` Success/Skipped/no-job, `1`
+/// Failure/Cancelled or a listener error, `2` env/config error before
+/// polling, `124` the deadline watchdog fired (takes precedence).
+pub(crate) async fn cmd_boot() -> i32 {
+  let masker = Arc::new(Mutex::new(SecretMasker::new()));
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:30]: As the finding itself concludes, this is the pre-existing pattern shared with run_cmd: the engine populates the masker from the job message's mask hints once a job is parsed, and pre-job lines (auth, session) cannot contain job secrets — none have reached the process yet. The JIT blob itself is never logged. No change.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+/// Current wall-clock time as epoch milliseconds, saturating instead of
+/// panicking on a `SystemTime` before the epoch or a `u128` overflow (both
+/// practically unreachable, but `expect`/`unwrap` are denied workspace-wide).
+fn now_epoch_ms() -> u64 {
+  let elapsed = std::time::SystemTime::now()
+    .duration_since(std::time::UNIX_EPOCH)
+    .unwrap_or_default();
+  u64::try_from(elapsed.as_millis()).unwrap_or(u64::MAX)
+}
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:230]: The finding's own analysis ends 'No finding here on async discipline' — agreed: SystemTime::now() is a cheap clock read, fine both in the sync prelude and inside the watchdog task. No change.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+    deadline_hit.store(true, Ordering::SeqCst);
+    tracing::error!("TOOLU_DEADLINE reached; cancelling the in-flight job");
+    cancel.cancel();
+
+    tokio::time::sleep(WATCHDOG_GRACE).await;
+    tracing::error!("deadline grace period elapsed without a clean exit; hard-exiting 124");
+    std::process::exit(124);
+  });
+}
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs]: The watchdog body cannot panic: now_epoch_ms is saturating (unwrap_or_default / unwrap_or(u64::MAX), boot_cmd.rs), sleep/select!/store/cancel don't panic, and the workspace lint table denies panic/unwrap/expect in non-test code (root Cargo.toml). Joining the handle would invert the design — main must not block on a task that sleeps until the deadline; on the graceful path the handle is dropped at process exit like any background task. No change.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+  }
+
+  exit_code_for(result)
+}
+
+/// Map the listener's outcome to an exit code (deadline handled by the
+/// caller, before this is consulted).
+fn exit_code_for(result: Result<Option<Conclusion>, RunnerError>) -> i32 {
+  match result {
+    Ok(None | Some(Conclusion::Success | Conclusion::Skipped)) => 0,
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:158]: Fixed in 684490b — the --help exit-0 line and docs/container-image.md now read 'job completed Success or Skipped, or the runner shut down (signal or session end) before any job was acquired', covering every Ok(None) path instead of only cancellation.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+    Ok(Some(Conclusion::Failure | Conclusion::Cancelled)) => 1,
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:159]: The finding's own trace concludes 'This matches the doc': watchdog-cancelled → deadline_hit → 124; externally cancelled → 1. That is the implemented and documented behavior. No change.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+    Err(e) => {
+      eprintln!("toolu-runner: boot: {e}");
+      1
+    },
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:163]: As the finding's own analysis states, 'the distinction between pre-poll errors (exit 2) and post-poll errors (exit 1) is correctly implemented': GitHubListener::new parse failures map to 2 in run_listener_once, listener.run failures to 1, matching the help text. No change.

File: crates/toolu-runner/src/main.rs [hunk 8]
```diff
@@ -33,6 +35,12 @@ async fn main() {
   #[cfg(debug_assertions)]
   cli::debug_assert_cli();
   let cli = Cli::parse();
+  // `boot` needs richer exit codes than every other subcommand's uniform
+  // Ok-is-0 / Err-is-2 (see `boot_cmd::cmd_boot`'s doc), so it is
+  // intercepted here rather than folded into `run`'s `Result` mapping.
+  if matches!(cli.command, Command::Boot(_)) {
+    std::process::exit(boot_cmd::cmd_boot().await);
+  }
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/main.rs:43]: There is no buffered writer to flush: shared::startup wires the fmt layers through RedactingMakeWriter/RedactingWriter with no tracing_appender non_blocking and no WorkerGuard anywhere in startup.rs, so every event is written synchronously at emit time. By the time cmd_boot returns its code, all lines are already on disk/stderr — process::exit drops nothing. Routing boot through run()'s uniform Ok-0/Err-2 mapping would conflate it with boot's four-code contract for no gain.

File: crates/toolu-runner/tests/boot_cmd_e2e_test.rs [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+  boot_fixtures::mount_job_lifecycle(&server, "sleep 60")
+    .await
+    .expect("mount the broker + run-service mocks");
+  let jit_config =
+    boot_fixtures::real_jit_config_b64(&server.uri()).expect("build a real-keypair jit config");
+
+  let now_ms = SystemTime::now()
+    .duration_since(UNIX_EPOCH)
+    .expect("system clock is after the epoch")
+    .as_millis();
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/boot_cmd_e2e_test.rs]: The finding's own analysis concludes 'this is fine. No finding' — allow-expect-in-tests covers #[test]/#[tokio::test] fns under tests/. No change.

File: crates/toolu-runner/tests/boot_cmd_e2e_test.rs [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    .output()
+    .expect("run binary");
+  let elapsed = started.elapsed();
+
+  let stderr = String::from_utf8_lossy(&output.stderr);
+  assert_eq!(
+    output.status.code(),
+    Some(124),
+    "expected exit 124 when the deadline watchdog fires mid-job, stderr: {stderr}"
+  );
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/boot_cmd_e2e_test.rs:169]: Fixed in 684490b — the watchdog test now asserts the broker received a /completejob call (broker_saw_completejob over server.received_requests()), pinning that the graceful cancel reported the job before exit rather than the 30s hard exit winning.

File: crates/toolu-runner/tests/boot_cmd_test.rs [hunk 11]
```diff
@@ -0,0 +1,139 @@
...
+    .env_remove("TOOLU_DEADLINE")
+    .output()
+    .expect("run binary");
+
+  assert_eq!(output.status.code(), Some(2), "expected exit 2");
+  let stderr = String::from_utf8_lossy(&output.stderr);
+  assert!(
+    stderr.contains("TOOLU_JITCONFIG"),
+    "stderr should name TOOLU_JITCONFIG: {stderr}"
+  );
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/boot_cmd_test.rs:60]: Fixed in 684490b — both missing-jitconfig tests now assert the full 'TOOLU_JITCONFIG is not set (or empty)' message rather than just the variable name.

File: crates/toolu-runner/tests/helpers/boot_fixtures.rs [hunk 12]
```diff
@@ -0,0 +1,156 @@
...
+  let rsa_params = protocol::RsaKeyParams {
+    exponent: be(public.e()),
+    modulus: be(public.n()),
+    d: be(private.d()),
+    p: be(p),
+    q: be(q),
+    // Recomputed by `parse_rsa_private_key` from (d, p, q); placeholders.
+    dp: BASE64.encode([0u8]),
+    dq: BASE64.encode([0u8]),
+    inverse_q: BASE64.encode([0u8]),
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/helpers/boot_fixtures.rs:44]: By design and already documented at the site: parse_rsa_private_key recomputes the CRT components from (d, p, q) — that is its contract (crates/protocol/src/auth.rs, the .NET RSACryptoServiceProvider params path). If it ever changed to trust dp/dq/inverse_q, these tests failing loudly is the desired signal, not silent breakage. The suggestion's text is the comment already present. No change.

File: crates/toolu-runner/tests/listener_smoke_test.rs [hunk 13]
```diff
@@ -129,3 +299,75 @@ async fn listener_rejects_invalid_jit_config() {
...
+  mount_auth_and_session(&server).await;
+
+  Mock::given(method("GET"))
+    .and(path("/message"))
+    .respond_with(ResponseTemplate::new(202))
+    .mount(&server)
+    .await;
+
+  let config = make_config();
+  let masker = Arc::new(std::sync::Mutex::new(SecretMasker::new()));
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs:325]: SecretMasker::new() is infallible — pub fn new() -> Self (crates/shared/src/secret_masker.rs). There is no Result to discard, and the suggested .expect() would not compile. No change.

File: crates/toolu-runner/tests/listener_smoke_test.rs [hunk 14]
```diff
@@ -129,3 +299,75 @@ async fn listener_rejects_invalid_jit_config() {
...
+#[tokio::test]
+async fn run_returns_the_completed_jobs_conclusion() {
+  let server = MockServer::start().await;
+  mount_auth_and_session(&server).await;
+  mount_job_lifecycle(&server)
+    .await
+    .expect("mount the broker + run-service mocks");
+
+  let (_dir, config) = unique_job_config().expect("build a fresh job-scoped RunnerConfig");
+  let masker = Arc::new(std::sync::Mutex::new(SecretMasker::new()));
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs:364]: Same as the line-320 thread: SecretMasker::new() is pub fn new() -> Self — infallible, no Result to handle; the suggestion would not compile. No change.

File: crates/toolu-runner/tests/listener_smoke_test.rs [hunk 15]
```diff
@@ -67,6 +71,172 @@ fn minimal_jit_config_b64(server_url_v2: &str, agent_id: i64) -> String {
...
+
+/// A `RunnerConfig` rooted at a fresh tempdir (both `data_dir` and
+/// `workspace_root` pre-created), so the real engine can actually run a job
+/// in it — unlike `make_config`'s uncreated, test-shared temp paths.
+fn unique_job_config() -> Result<(tempfile::TempDir, RunnerConfig), String> {
+  let dir = tempfile::tempdir().map_err(|e| format!("tempdir: {e}"))?;
+  let workspace_root = dir.path().join("work");
+  let data_dir = dir.path().join("data");
+  std::fs::create_dir_all(&workspace_root).map_err(|e| format!("create workspace_root: {e}"))?;
+  std::fs::create_dir_all(&data_dir).map_err(|e| format!("create data_dir: {e}"))?;
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs]: Test-setup precedent across the suite: std::fs::create_dir_all appears in 10+ integration test files including #[tokio::test] fns (journal_writer_test.rs:39, config_resolve_test.rs, gh_compat_*, …). This is cold-path test scaffolding, not production code; wrapping it in spawn_blocking would diverge from every sibling test for no observable benefit. No change.

File: crates/toolu-runner/tests/listener_smoke_test.rs [hunk 16]
```diff
@@ -67,6 +71,172 @@ fn minimal_jit_config_b64(server_url_v2: &str, agent_id: i64) -> String {
...
+  let rsa_params = protocol::RsaKeyParams {
+    exponent: be(public.e()),
+    modulus: be(public.n()),
+    d: be(private.d()),
+    p: be(p),
+    q: be(q),
+    // Recomputed by `parse_rsa_private_key` from (d, p, q); placeholders.
+    dp: BASE64.encode([0u8]),
+    dq: BASE64.encode([0u8]),
+    inverse_q: BASE64.encode([0u8]),
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs:102]: The suggestion is byte-identical to the comment already in the file; the placeholders are the documented contract (parse_rsa_private_key recomputes the CRT components from d, p, q). No change.

File: crates/toolu-runner/src/boot_cmd.rs [hunk 17]
```diff
@@ -0,0 +1,238 @@
...
+    .map_err(|e| format!("HTTP client: {e}"))
+}
+
+/// Spawn the deadline watchdog: sleep until `deadline_ms` (standing down
+/// early if `cancel` fires first — a plain SIGINT/SIGTERM is not a deadline
+/// hit), then fire `cancel` for a graceful shutdown (the job reports
+/// `Cancelled` to GitHub) and set `deadline_hit`. If the graceful path
+/// hasn't exited the process within [`WATCHDOG_GRACE`], hard-exit 124.
+fn spawn_watchdog(deadline_ms: u64, cancel: CancellationToken, deadline_hit: Arc<AtomicBool>) {
+  tokio::spawn(async move {
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:281]: Recurrence of the thread refused last round (boot_cmd.rs:221, now resolved): the watchdog body has no reachable panic — now_epoch_ms is saturating (unwrap_or_default / unwrap_or(u64::MAX)), tokio::time::sleep and cancel() don't panic, and the workspace denies panic/unwrap/expect in non-test code. There is no caller that can await the handle without blocking main on a task that sleeps until the deadline; on the normal path the process exits first by design. Standing disagreement — escalating to the PR author rather than looping.

File: crates/toolu-runner/tests/listener_smoke_test.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs]: Recurrence of the thread refused last round (listener_smoke_test.rs:229, now resolved): std::fs::create_dir_all in test setup is the established pattern across 10+ integration test files in this suite (journal_writer_test.rs:39, config_resolve_test.rs, gh_compat_*). The finding itself rates it 'minor... worth noting for consistency' — and consistency here means matching the sibling tests, not diverging from them. Standing disagreement — escalating to the PR author rather than looping.

File: crates/toolu-runner/src/boot_cmd.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:311]: Reversed — fixed in 6cf9eb0. The PR author asked for every finding to be implemented rather than argued, so the hard exit is no longer the only thing standing between a deadline and the process: `cmd_boot` now owns the watchdog's `JoinHandle` and aborts + joins it once the listener returns, so the grace-period sleep can never outlive the job, and `std::process::exit(124)` flushes stderr before it runs (the `_diag` sink is `tracing_appender::rolling`, synchronous, so it was already durable). Pinned by boot_cmd_e2e_test.rs::boot_stands_the_watchdog_down_when_the_job_finishes_first (0ca0543), and the mid-job test's bound is tightened 40s -> 25s so it can no longer pass via the hard-exit path.

File: crates/toolu-runner/src/boot_cmd.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:30]: Reversed — fixed in 6cf9eb0. Rather than defend the empty-masker pattern, `boot` now registers the raw `TOOLU_JITCONFIG` blob with the masker immediately after `resolve_boot_inputs`, before anything else can log: the blob carries the runner's RSA private key, so it no longer depends on the engine's job-message mask hints for its own credential. Pinned in 0ca0543 — the success e2e asserts the blob appears verbatim in neither `_diag/runner.log` nor stderr, so deleting the registration now fails the suite.

File: crates/toolu-runner/src/boot_cmd.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/src/boot_cmd.rs:281]: Reversed — fixed in 6cf9eb0. `spawn_watchdog` now returns its `JoinHandle`; `cmd_boot` aborts and joins it after `run_listener_once`, logging only a genuine panic (abort yields `is_cancelled`, not `is_panic`). This also closes the concern the previous reply raised about awaiting a task that sleeps until the deadline — the join happens after the lifecycle, so it returns immediately. New test in 0ca0543 pins it.

File: crates/toolu-runner/tests/boot_cmd_e2e_test.rs (hunk 9 above)
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/boot_cmd_e2e_test.rs]: Fixed in 6cf9eb0 — the `expect` is gone, replaced with `unwrap_or_default()` behind a `deadline_in(offset)` helper now shared by both deadline tests.

File: crates/toolu-runner/tests/listener_smoke_test.rs (hunk 15 above)
[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs]: Reversed — fixed in 6cf9eb0. `unique_job_config` is now `async` and wraps both `create_dir_all` calls in `tokio::task::spawn_blocking`, and its error type moved from `String` to `Box<dyn std::error::Error>` so the source chain survives (the other finding on this helper). The sole call site awaits it. Sibling-test consistency was the prior argument; the author's call is to follow the guideline instead.

[the remark this answers is not in the document]
Comment (Falconiere) [reply crates/toolu-runner/tests/listener_smoke_test.rs]: Reversed — fixed in 6cf9eb0, same change as the sibling thread on this helper: `unique_job_config` is `async` with both `create_dir_all` calls inside `tokio::task::spawn_blocking`, and the `String` error type is now `Box<dyn std::error::Error>`. Full suite green (654 tests).

File: docs/container-image.md [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+a JIT config is single-use by design, and the providers destroy the
+instance after the job (`restart.policy: "no"` + `auto_destroy` on
+Fly; deadline-driven teardown on Namespace).
+
+When a deadline is set, a watchdog fires at that instant: it cancels
+the in-flight job gracefully (the job reports Cancelled to GitHub) and
+hard-exits after a 30-second grace period. The graceful path normally
+wins — once the listener returns, `boot` stands the watchdog down and
+exits `124` through its own return path, so the hard exit is reached
+only when shutdown itself hangs.
```
[the remark this answers is not in the document]
Comment (Falconiere) [reply docs/container-image.md:38]: Fixed in 04d1bf2 — docs/container-image.md now names `std::process::exit(124)` explicitly and spells out the trade: it runs no destructors, so the journal writer, a step's log upload, and the `/completejob` report are abandoned mid-write. What survives is stated too (stderr is hand-flushed immediately before the exit; the `_diag` sink is `tracing_appender::rolling`, synchronous with no background worker, so what was already logged is on disk), along with the consequence a reader actually cares about — a job on this path may sit in-progress in the GitHub UI until GitHub times it out.
